xcdat/dataset.py

```diff
@@ -35,13 +37,97 @@
 ]
 
 
-def open_dataset(
+def open(
     path: str,
     data_var: Optional[str] = None,
     add_bounds: bool = True,
     decode_times: bool = True,
     center_times: bool = False,
     lon_orient: Optional[Tuple[float, float]] = None,
+    **kwargs,
+) -> xr.Dataset:
+    """
+    Documentation to be written when this PR is
+    further along...
+    """
+
+    # inspect path attributes and use either open_dataset
+    # or open_mfdataset as appropriate
+    #   * list (open list of files with open_mfdataset)
+    #   * pathlib.Path (open pathlib.Path with open_mfdataset)
+    #   * deprecated CDAT xml file type
+    #   * directory (opens with open_dataset or open_mfdataset
+    #                depending in number of files)
+    #   * wildcard input (opens data with open_mfdataset)
+
+    # Note: if a directory / xml files / wildcard is provided, we
+    #       will simply glob the entire directory with open_mfdataset.
+    if (type(path) == list) | (type(path) == pathlib.Path):
+        mfdataset = True  # lists/pathlib.paths are handled by open_mfdataset
+    elif path.split(".")[-1] == "xml":
+        # parse the xml file for the directory attribute
+        # then use open_mfdataset with a wildcard search
+        # for all files in that directory
+        tree = ET.parse(path)
+        root = tree.getroot()
+        d = root.attrib["directory"]
+        path = d + "/*"
+        mfdataset = True
+    elif os.path.isdir(path):
+        # if the path is a directory, check the number of files
+        # in that directory. If one file is present, use
+        # open_dataset. If multiple files are present, use
+        # open_mfdataset with a wildcard for all files in that
+        # directory. If no files are present, throw a ValueError.
+        fileList = os.listdir(path)
+        if len(fileList) == 0:
+            ValueError("The supplied directory has no files to open.")
+        elif len(fileList) == 1:
+            mfdataset = False
+            path = path + "/" + fileList[0]
+        else:
+            mfdataset = True
+            path = path + "/*"
+    elif "*" in path:
+        # if path includes a wildcard operator, pass the path
+        # to open_mfdataset
+        mfdataset = True
+    else:
+        # if none of these criteria are fulfilled, try open_dataset
+        mfdataset = False
+
+    # use open_dataset or open_mfdataset depending on provided path
+    if mfdataset:
+        ds = open_mfdataset(
+            path,
+            data_var=data_var,
+            add_bounds=add_bounds,
+            decode_times=decode_times,
+            center_times=center_times,
+            lon_orient=lon_orient,
+            **kwargs,
+        )
+    else:
+        ds = open_dataset(
+            path,
+            data_var=data_var,
+            add_bounds=add_bounds,
+            decode_times=decode_times,
+            center_times=center_times,
+            lon_orient=lon_orient,
+            **kwargs,
+        )
```

Comment:
I noticed this logic for determining whether to use `open_dataset()` or `open_mfdataset()`.

**`open_mfdataset()` supports opening a single dataset file and multiple dataset files.** The main difference (that I'm aware of) with opening a single dataset with `open_dataset()` vs. `open_mfdataset()` is that data variables are automatically Dask arrays with `open_mfdataset()`. 

I'm not sure of pros/cons, implications (e.g. performance), etc. of using `open_dataset()` vs. `open_mfdataset()` for a single dataset file, but **I'd probably lean towards always using `open_mfdataset()` if you're looping over models or lots of directories.**

Reply:
When I loop over models I currently use `open_mfdataset` universally: `xc.open_mfdataset(dpath + '*.nc')`. `xc.open_dataset()` used to be faster than `xc.open_mfdataset()`, though this may have been addressed a few months ago (I remember discussing this, perhaps as part of the large PR that changes the logic in dealing with coords/bounds and I think it may have been addressed). 

I just checked for one dataset and they were close (24 versus 21 ms), though I should look to see if both functions are close in speed across a broader range of datasets.

Reply:
I'd be fine with this, but ideally `open_mfdataset()` could also handle a string [`directory`](https://github.com/xCDAT/xcdat/blob/3913ea2cba91a24d099686072832cebf62ac4911/xcdat/dataset.py#L76). I think this would handle all cases in this PR. 

I think I personally would then just use `open_mfdataset()`. 

The case where `open()` would still be useful is if `open_dataset()` is faster than `open_mfdataset()`; `open()` would be faster  / better by intelligently using `open_mfdataset()` only when necessary.

Reply:
> The case where `open()` would still be useful is if `open_dataset()` is faster than `open_mfdataset()`; `open()` would be faster / better by intelligently using `open_mfdataset()` only when necessary.

I found some xarray GH issues related to the slow performance of `open_mfdataset()`, which are **mostly be related to [coordinate compatibility checks when concatenating files](https://github.com/pydata/xarray/issues/1385#issuecomment-561920115) and [decoding time coordinates](https://github.com/pydata/xarray/issues/1385#issuecomment-370064483) with more than 1 dataset**. There are [workarounds](https://docs.xarray.dev/en/stable/user-guide/io.html#reading-multi-file-datasets) for users if they find `open_mfdataset()` too slow.

**For our specific case involving opening only 1 dataset** with `open_dataset()` vs. `open_mfdataset()`, I don't think the performance difference is that significant because coordinate concatenation isn't necessary like it is with multiple datasets. **I seems safe to extend `open_mfdataset()`, but some performance metrics would be nice to confirm this.** 

> I'd be fine with this, but ideally `open_mfdataset()` could also handle a string [`directory`](https://github.com/xCDAT/xcdat/blob/3913ea2cba91a24d099686072832cebf62ac4911/xcdat/dataset.py#L76). I think this would handle all cases in this PR.

I'll take a look at your implementation and suggest some changes to extend `open_mfdataset()`.